server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSReceiveOptionsCheckUtility.java

```diff
@@ -107,12 +110,64 @@ private void checkPartn(CICSParser.Cics_receive_partnContext ctx) {
 
   private void checkMap(CICSParser.Cics_receive_mapContext ctx) {
     if (ctx.FROM().isEmpty()) checkHasIllegalOptions(ctx.LENGTH(), "LENGTH without FROM");
+    if (!checkMapHasLiteral(ctx)) {
+      checkHasMandatoryOptions(
+          ctx.cics_into_set().stream()
+              .map(CICSParser.Cics_into_setContext::INTO)
+              .collect(Collectors.toList()),
+          ctx,
+          "INTO when specifying MAP param without literal");
+    }
     checkIntoSet(ctx.cics_into_set());
+    checkHasMutuallyExclusiveOptions("TERMINAL or FROM", ctx.TERMINAL(), ctx.FROM());
   }
 
   private void checkMapMappingDev(CICSParser.Cics_receive_map_mappingdevContext ctx) {
     checkHasMandatoryOptions(ctx.FROM(), ctx, "FROM");
-    checkIntoSet(ctx.cics_into_set());
+    if (!checkMapHasLiteral(ctx)) {
+      checkHasMandatoryOptions(
+          ctx.cics_into_set().stream()
+              .map(CICSParser.Cics_into_setContext::INTO)
+              .collect(Collectors.toList()),
+          ctx,
+          "INTO when specifying MAP param without literal");
+    }
+  }
+
+  private boolean checkMapHasLiteral(ParserRuleContext ctx) {
+    if (ctx.children == null) return false;
+    boolean hasLiteral = false;
+    for (int index = 0; index < ctx.children.size(); index++) {
+      ParseTree item = ctx.children.get(index);
+      if (TerminalNode.class.isAssignableFrom(item.getClass())) {
+        TerminalNode node = (TerminalNode) item;
+        if (node.getSymbol().getType() == CICSParser.MAP && index + 1 < ctx.children.size() - 1) {
```

Comment:
```suggestion
        if (node.getSymbol().getType() == CICSParser.MAP && index + 1 < ctx.children.size()) {
```

Reply:
or in this case you could iterate to `ctx.children.size() - 1` in the outer loop.